### PR TITLE
Builder based initializer for `DictionaryExpr`

### PR DESCRIPTION
**Pull request by evnik**
Allow to use `ArrayType` and `DictionaryType` as `PatternBinding.typeAnnotation` argument

### CODE REVIEW

**Review by ahoppen — COMMENTED** [review]
This looks good but I think there’s a slight mismatch between PR title, description in code. AFAICT
- The description mentions “Allow to use ArrayType and DictionaryType as PatternBinding.typeAnnotation argument” covers all changes except DictionaryExprConvenienceInitializers.swift. These changes are looking great. ✅ 
- The title “Builder based initializer for DictionaryExpr” covers the changes in DictionaryExprConvenienceInitializers.swift. It would be good to have test coverage for those before merging. Maybe we can just cherry-pick the test case from https://github.com/apple/swift-syntax/pull/454.

**Comment by evnik** [comment]
My intention was to avoid adding trailing commas, so I won't need to remove them in #454, but that's fine as well

**Review by ahoppen — APPROVED** [review]
Thanks. Looks good 👍 

I just prefer to keep the PRs isolated so we can’t forget to add the test case if for some reason we should decide to take a different direction on the other PR. It also makes for a nicer git history where you can see the test cases that were added for a particular code change.

**Comment by ahoppen** [comment]
@swift-ci Please test

**Pull request merged**
